pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1151,34 +866,32 @@ protected IdealState ensureAllPartitionsConsuming(final TableConfig tableConfig,
-            LOGGER.info("Found smallest offset {} for table {} for partition {}", startOffset, tableNameWithType,
-                partition);
-            startOffset = getBetterStartOffsetIfNeeded(tableNameWithType, partition, segmentName, startOffset,
-                newLLCSegmentName.getSequenceNumber());
+          if (isAllInstancesInState(instanceStateMap, RealtimeSegmentOnlineOfflineStateModel.OFFLINE)) {
+            LOGGER.info("Repairing segment: {} which is OFFLINE for all instances in IdealState", latestSegmentName);
+
+            // Create a new segment to re-consume from the previous start offset
+            LLCSegmentName newLLCSegmentName = getNextLLCSegmentName(latestLLCSegmentName, currentTimeMs);
+            long startOffset = latestSegmentZKMetadata.getStartOffset();
+            // Start offset must be higher than the start offset of the stream
+            long partitionStartOffset =
+                getPartitionOffset(streamConfig, OffsetCriteria.SMALLEST_OFFSET_CRITERIA, partitionId);
+            if (partitionStartOffset > startOffset) {
+              LOGGER.error("Data lost from offset: {} to: {} for partition: {} of table: {}", startOffset,
+                  partitionStartOffset, partitionId, realtimeTableName);
+              _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.LLC_KAFKA_DATA_LOSS, 1L);
+              _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.LLC_STREAM_DATA_LOSS, 1L);
+              startOffset = partitionStartOffset;
+            }
 
             CommittingSegmentDescriptor committingSegmentDescriptor =
-                new CommittingSegmentDescriptor(segmentId, startOffset, 0);
-
-            boolean success =
-                createNewSegmentMetadataZNRecord(tableConfig, segmentName, newLLCSegmentName, partitionAssignment,
-                    committingSegmentDescriptor, false);
-
-            // creation of segment metadata could fail due to lost leadership or an unsuccessful write to property store
-            // in such a case, we will exclude the segment from ideal state update and let the next iteration of validation manager fix it
-            if (success) {
-              consumingSegments.add(newLLCSegmentName.getSegmentName());
-            }
+                new CommittingSegmentDescriptor(latestSegmentName, startOffset, 0);
+            createNewSegmentZKMetadata(tableConfig, streamConfig, newLLCSegmentName, currentTimeMs,
+                committingSegmentDescriptor, latestSegmentZKMetadata, instancePartitions, numPartitions, numReplicas);
+            String newSegmentName = newLLCSegmentName.getSegmentName();
+            updateInstanceStatesForNewConsumingSegment(instanceStatesMap, null, newSegmentName, segmentAssignment,
+                instancePartitionsMap);
+          } else {
+            LOGGER.error("Got unexpected instance state map: {} for segment: {}", instanceStateMap, latestSegmentName);
```

Comment:
Why is this unexpected state?

Reply:
This is for debugging purpose to log unexpected state that is not properly handled. E.g. `all in ONLINE state` scenario is not handled properly in RetentionManager

Reply:
@npawar  and I discussed and came to the conclusion that we were being conservative when we handle the all ONLINE case. Can you please add a comment to that effect?

Reply:
Added more comment explaining what needs to be handled when getting this error message.